### DISCUSSION TITLE
Best practice for editing data

### DISCUSSION THREAD

**Discussion by dennis-8** [TanStack/query#3870] [Q&A]
Hi,

I'm confused by what is the best way to edit data using React Query.

For example, I have a `useState` variable:
`const [person, setPerson] = useState<Person>({name: ""});`

It's used when I edit a person's name: 
`<input value={person.name} onChange={handleChange}/>`

Where does `useQuery` fit here?

For example, I fetch the user with id 1: `const { data } = useQuery(["users", {id: 1}], fetchUsers);`

What is the correct way to edit user with id 1? I can pass `data` to a component which has the form for editing name.
Usually the `data` would be the default value for useState, something like this: `useState<Person>(data)`;

What if I start to edit user 1, but somebody else changes the name of user 1 and my app makes a new fetch after that, before I save my changes.

How this should be handled?

**Comment by clicktodev** [comment]
You can do something like this if I understood you correctly.

This will refetch the user everytime the user name is changed in react.

In order to fetch the users after the user is updated on the server, you will have to use a mutation and then refetch the users when the mutation is successful. Below is a simple example without any mutations.

https://codesandbox.io/s/dry-currying-9n9tpe?file=/src/index.js:1169-1175
```js
function usePosts(person) {
  return useQuery(["posts"], async () => {
    const { data } = await axios.get(
      "https://jsonplaceholder.typicode.com/posts"
    );
    return data;
  });
}

function Posts({ setPostId }) {
  const queryClient = useQueryClient();
  const [person, setPerson] = useState("default");

  const { status, data, error, isFetching } = usePosts(person);

  useEffect(() => {
    queryClient.invalidateQueries(["posts"]);
  }, [person, queryClient]);

  return (
    <div>
      <button onClick={() => setPerson(Math.random())}>
        change person: {person}
      </button>
      <h1>Posts</h1>
      <div>
        {status === "loading" ? (
          "Loading..."
        ) : status === "error" ? (
          <span>Error: {error.message}</span>
        ) : (
          <>
            <div>
              {data.map((post) => (
                <p key={post.id}>
                  <a onClick={() => setPostId(post.id)} href="#">
                    {post.title}
                  </a>
                </p>
              ))}
            </div>
            <div>{isFetching ? "Background Updating..." : " "}</div>
          </>
        )}
      </div>
    </div>
  );
}
```

**Reply by clicktodev** [reply]
I will be happy to provide an example with this flow:

1: update user state in react
2: update the user on the server
3: refetch the updated list of users

**Comment by dennis-8** [comment]
Thank you,

I'd be interested in an example when you have multiple posts on the page and you edit them without leaving the page.

For example, you start to edit post #1, but some other user already updated it.
You open another tab without saving your changes then return to the tab where you plan to save your changes.
React-query will make a request to get latest changes and will return updated post #1.

What should be done in this case? Lose your changes and show the latest changes from server?

**Comment by TkDodo** [comment]
I did go into a bunch of options in this blog post: https://tkdodo.eu/blog/react-query-and-forms

**Reply by clicktodev** [reply]
@dennis-8 definitely follow the blog @TkDodo wrote. He always delivers useful content.
